pkg/cmd/pulumi/plugin_install.go

```diff
@@ -62,161 +59,204 @@ func newPluginInstallCmd() *cobra.Command {
+		"checksum", "", "The expected SHA256 checksum for the plugin archive")
+
+	return cmd
+}
+
+type pluginInstallCmd struct {
+	serverURL string
+	exact     bool
+	file      string
+	reinstall bool
+	checksum  string
+
+	d     diag.Sink
+	e     env.Env
+	color colors.Colorization
+}
+
+func (cmd *pluginInstallCmd) Run(ctx context.Context, args []string) error {
+	if cmd.e == nil {
+		cmd.e = env.Global()
+	}
+	if cmd.d == nil {
+		cmd.d = cmdutil.Diag()
+	}
+	if cmd.color == "" {
+		cmd.color = cmdutil.GetGlobalColorization()
+	}
+
+	// Parse the kind, name, and version, if specified.
+	var installs []workspace.PluginSpec
+	if len(args) > 0 {
+		if !workspace.IsPluginKind(args[0]) {
+			return fmt.Errorf("unrecognized plugin kind: %s", args[0])
+		} else if len(args) < 2 {
+			return errors.New("missing plugin name argument")
+		}
+
+		var version *semver.Version
+		if len(args) == 3 {
+			parsedVersion, err := semver.ParseTolerant(args[2])
+			version = &parsedVersion
+			if err != nil {
+				return fmt.Errorf("invalid plugin semver: %w", err)
 			}
+		}
+		if len(args) < 3 && cmd.file != "" {
+			return errors.New("missing plugin version argument, this is required if installing from a file")
+		}
 
-			// Parse the kind, name, and version, if specified.
-			var installs []workspace.PluginSpec
-			if len(args) > 0 {
-				if !workspace.IsPluginKind(args[0]) {
-					return fmt.Errorf("unrecognized plugin kind: %s", args[0])
-				} else if len(args) < 2 {
-					return errors.New("missing plugin name argument")
-				}
+		var checksums map[string][]byte
+		if cmd.checksum != "" {
+			checksumBytes, err := hex.DecodeString(cmd.checksum)
+			if err != nil {
+				return fmt.Errorf("--checksum was not a valid hex string: %w", err)
+			}
+			checksums = map[string][]byte{
+				runtime.GOOS + "-" + runtime.GOARCH: checksumBytes,
+			}
+		}
 
-				var version *semver.Version
-				if len(args) == 3 {
-					parsedVersion, err := semver.ParseTolerant(args[2])
-					version = &parsedVersion
-					if err != nil {
-						return fmt.Errorf("invalid plugin semver: %w", err)
-					}
-				}
-				if len(args) < 3 && file != "" {
-					return errors.New("missing plugin version argument, this is required if installing from a file")
-				}
+		pluginSpec := workspace.PluginSpec{
+			Kind:              workspace.PluginKind(args[0]),
+			Name:              args[1],
+			Version:           version,
+			PluginDownloadURL: cmd.serverURL, // If empty, will use default plugin source.
+			Checksums:         checksums,
+		}
 
-				var checksums map[string][]byte
-				if checksum != "" {
-					checksumBytes, err := hex.DecodeString(checksum)
-					if err != nil {
-						return fmt.Errorf("--checksum was not a valid hex string: %w", err)
-					}
-					checksums = map[string][]byte{
-						runtime.GOOS + "-" + runtime.GOARCH: checksumBytes,
-					}
-				}
+		// Bundled plugins are generally not installable with this command. They are expected to be
```

Comment:
This is the only real change here, the rest is just indent change from putting this in the Run method.